sui-execution/latest/sui-verifier/src/meter.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_bytecode_verifier::{
+    meter::{Meter, Scope},
+    VerifierConfig,
+};
+use move_core_types::vm_status::StatusCode;
+
+struct SuiVerifierMeterBounds {
+    name: String,
+    ticks: u128,
+    max_ticks: Option<u128>,
+}
+
+impl SuiVerifierMeterBounds {
+    fn add(&mut self, ticks: u128) -> PartialVMResult<()> {
+        if let Some(max_ticks) = self.max_ticks {
+            let new_ticks = self.ticks.saturating_add(ticks);
+            if new_ticks > max_ticks {
+                return Err(PartialVMError::new(StatusCode::CONSTRAINT_NOT_SATISFIED)
+                    .with_message(format!(
+                        "program too complex. Ticks exceeded `{}` will exceed limits: `{} current + {} new > {} max`)",
+                        self.name, self.ticks, ticks, max_ticks
+                    )));
+            }
+            self.ticks = new_ticks;
+        }
+        Ok(())
+    }
+}
```

Comment:
Can we re-use `Bounds` in the move implementation?

Reply:
With the current impl, yes. But I'm making some subsequent changes e.g no optional `max_ticks`. This way one cannot accidentally have unbounded metering.
However since the code is shared, between us and others, I don't want to force other users to change their meter.